libs/wingc/src/jsify.rs

```diff
@@ -1149,3 +1168,178 @@ fn is_mutable_collection(expression: &Expr) -> bool {
+		(self.references, self.diagnostics)
+	}
+}
+
+impl<'ast> Visit<'ast> for FieldReferenceVisitor<'_> {
+	fn visit_reference(&mut self, node: &'ast Reference) {
+
+		if self.collect_inflight_refs(node) {
+			return;
+		}
+
+		visit::visit_reference(self, node);
+	}
+}
+
+impl<'a> FieldReferenceVisitor<'a> {
+
+	/// If node is a reference to a field of "this", check that it references a legal field and add it
+	/// to the list of references. Returns true if no additional processing is needed or false if we
+	/// need to visit the node's children.
+	fn collect_inflight_refs(&mut self, node: &Reference) -> bool {
+		let (object, property) = match node {
+			Reference::InstanceMember { object, property } => (object, property),
+			_ => return false,
+		};
+
+		let r = match &object.kind {
+			ExprKind::Reference(r) => r,
+			_ => return false
+		};
+
+		let s = match r {
+			Reference::Identifier(s) => s,
+			_ => {
+				self.path.insert(0, property.name.clone());
+				return false
+			}
+		};
+
+		if s.name != "this" {
+			return false
+		}
+
+		// resolve the type of the resource by looking up "this" in our environment
+		let resource_type = self.env
+			.lookup(s, None).expect("'this' to be found")
+			.as_variable().expect("'this' to be a variable")
+			.type_;
+
+		let r = match &*resource_type {
+			Type::Resource(r) => r,
+			_ => panic!("'this' is not a resource"),
+		};
+
+		// lookup our field in the class environment
+		let field_kind = r.env
+			.lookup(property, None).expect("unable to find field in resource env")
+			.as_variable().expect("field is not a variable");
+
+		// we only care about preflight fields (inflight fields are normal references)
+		if field_kind.flight == Phase::Preflight {
+
+			// don't allow capturing reassignable (`var`) fields.
+			if field_kind.reassignable {
+				self.diagnostics.insert(Diagnostic {
+					level: DiagnosticLevel::Error,
+					message: format!(
+						"Unable to reference \"this.{}\" from inflight method \"{}\" because it is reassignable (\"var\")",
+						property.name, 
+						self.method_name.name,
+					),
+					span: Some(property.span.clone()),
+				});
+
+				return true
+			}
+
+			// check if the type is capturable (resource, primitive or immutable collection)
+			if ! field_kind.type_.is_capturable() {
+				self.diagnostics.insert(Diagnostic {
+					level: DiagnosticLevel::Error,
+					message: format!(
+						"Unable to reference \"this.{}\" from inflight method \"{}\" because type {} is not capturable",
+						property.name, 
+						self.method_name.name,
+						field_kind.type_,
+					),
+					span: Some(property.span.clone()),
+				});
+
+				return true
+			}
+
+			debug!("field \"this.{}\" is referenced from inflight method {}", property.name, self.method_name.name);
+
+			// get/create an "ops" set for this field
+			let ops = self.references.entry(property.name.clone()).or_default();
+
+			if ! self.path.is_empty() {
+				let op = self.path[0].clone();
```

Comment:
I got a little mixed up with this, why do we need only the top of the `path`?
And What happens when we access an inner field: `this.some_field.inner_field.do()`?

Reply:
See [this test](https://github.com/winglang/wing/blob/eladb/infer-refs/examples/tests/invalid/inflight_ref_resource_sub_field.w).

Not ideal, but it's a start.

Reply:
OK, just saw this. Next time hit "unresolve conversation" if you think there's still something to cover...

Reply:
See https://github.com/winglang/wing/pull/1682/commits/b4a30ffad00e0dad97d19ac14a496463408ed036 - basically failing this should be covered in the next PR.